drift-javadoc/src/main/java/io/airlift/drift/javadoc/JavadocProcessor.java

```diff
@@ -0,0 +1,196 @@
+import javax.lang.model.element.VariableElement;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Elements;
+import javax.tools.Diagnostic;
+
+import java.io.IOException;
+import java.io.Writer;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static io.airlift.drift.javadoc.ThriftAnnotations.META_SUFFIX;
+import static io.airlift.drift.javadoc.ThriftAnnotations.THRIFT_FIELD;
+import static io.airlift.drift.javadoc.ThriftAnnotations.THRIFT_METHOD;
+import static java.lang.String.format;
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+import static javax.lang.model.element.ElementKind.ENUM_CONSTANT;
+import static javax.tools.Diagnostic.Kind.ERROR;
+import static javax.tools.Diagnostic.Kind.NOTE;
+import static javax.tools.Diagnostic.Kind.WARNING;
+
+@SupportedAnnotationTypes({
+        ThriftAnnotations.THRIFT_ENUM,
+        ThriftAnnotations.THRIFT_SERVICE,
+        ThriftAnnotations.THRIFT_STRUCT})
+@SupportedSourceVersion(SourceVersion.RELEASE_8)
+public class JavadocProcessor
+        extends AbstractProcessor
+{
+    @Override
+    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment round)
+    {
+        for (TypeElement annotation : annotations) {
+            for (Element element : round.getElementsAnnotatedWith(annotation)) {
+                if (element instanceof TypeElement) {
+                    log(NOTE, "Extracting Javadoc metadata for " + element);
+                    extract((TypeElement) element);
+                }
+            }
+        }
+        return false;
+    }
+
+    private void extract(TypeElement typeElement)
+    {
+        if (typeElement.getQualifiedName().toString().endsWith(META_SUFFIX)) {
+            return;
+        }
+
+        switch (typeElement.getKind()) {
+            case CLASS:
+            case INTERFACE:
+            case ENUM:
+                break;
+            default:
+                log(WARNING, format("Non-class was annotated: %s %s", typeElement.getKind(), typeElement));
+                return;
+        }
+
+        List<String> serviceDocumentation = getComment(typeElement);
+
+        String packageName = elements().getPackageOf(typeElement).getQualifiedName().toString();
+        String className = getClassName(typeElement);
+
+        TypeSpec.Builder typeSpec = TypeSpec.classBuilder(className + META_SUFFIX)
+                .addAnnotation(documentationAnnotation(serviceDocumentation));
+
+        // offset auto-generated order numbers so they don't collide with hand-written ones
+        AtomicInteger nextOrder = new AtomicInteger(10000);
+
+        for (Element element : elements().getAllMembers(typeElement)) {
+            String name = element.getSimpleName().toString();
+            List<String> comment = getComment(element);
+            if ((element instanceof ExecutableElement) &&
```

Comment:
I tried that and the line length combined with the `&&` and `||` makes it harder to read